yare-engine/src/main/java/com/sabre/oss/yare/engine/executor/runtime/predicate/PredicateContext.java

```diff
@@ -36,13 +41,39 @@
     private final Object result;
     private final Map<String, Object> facts;
     private final Map<String, Object> attributes;
+    private final EngineController engineController;
+    private final EnumMap<ReservedIdentifier, Object> reservedIdentifiers;
 
     // Do not pass merged maps due to performance implications.
-    public PredicateContext(String ruleId, Object result, Map<String, Object> facts, Map<String, Object> attributes) {
+    public PredicateContext(String ruleId, Object result, Map<String, Object> facts, Map<String, Object> attributes, EngineController engineController) {
         this.ruleId = ruleId;
         this.result = result;
         this.facts = facts;
         this.attributes = attributes;
+        this.engineController = engineController;
+        this.reservedIdentifiers = createResolveIdentifiers();
+    }
+
+    private EnumMap<ReservedIdentifier, Object> createResolveIdentifiers() {
```

Comment:
I think this is overcomplicated. Using enum here gives absolutely no advantage. Actually adding one more if would be much more obvious and clear. I am not agains on map approach though. But it would be much simpler if it would be just `Map<String, Object>`. That way instead of this loop and adding `ReservedIdentifier` enum it would be just something like this:
```java
Map<String, Object> reserved - new HashMap<>();
reserved.put(CTX, result);
reserved.put(RULE_NAME, ruleId);
reserved.put(ENGINE_CONTROLLER, engineController);
return reserved;
````

Reply:
We will replace enum with string according your idea.

Reply:
Done.